include/openthread/tcp.h

```diff
@@ -0,0 +1,521 @@
+
+/**
+ * Binds the active socket to an IP address and port.
+ *
+ * @param[in]  aInstance   A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket     A pointer to the active TCP socket structure to bind.
+ * @param[in]  aSockName   The address and port to which to bind this socket.
+ *
+ * @retval OT_ERROR_NONE    Successfully bound the active TCP socket.
+ * @retval OT_ERROR_FAILED  Failed to bind the active TCP socket.
+ */
+otError otTcpActiveBind(otInstance *aInstance, otTcpActiveSocket *aSocket, const otSockAddr *aSockName);
+
+/**
+ * This enumeration defines flags passed to @p otTcpActiveConnect.
+ */
+enum {
+    OT_TCP_ACTIVE_CONNECT_FAST_OPEN =    1 << 0,
+    OT_TCP_ACTIVE_CONNECT_NO_FAST_OPEN = 1 << 1,
+};
+
+/**
+ * Records the remote host and port for this connection. If TCP Fast Open is
+ * used (it can be enabled or disabled using @p aFlags), then the TCP
+ * connection establishment handshake is initiated immediately; otherwise, it
+ * is initiated when data is first added to the send buffer.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket    A pointer to the active TCP socket structure to connect.
+ * @param[in]  aSockName  The IP address and port of the host to which to connect.
+ * @param[in]  aFlags     Flags specifying options for this operation (see enumeration above).
+ *
+ * @retval OT_ERROR_NONE    Successfully completed the operation.
+ * @retval OT_ERROR_FAILED  Failed to complete the operation.
+ */
+otError otTcpActiveConnect(otInstance *aInstance, otTcpActiveSocket *aSocket, const otSockAddr *aSockName, uint32_t aFlags);
+
+/**
+ * This enumeration defines flags passed to @p otTcpActiveSendByReference..
+ */
+enum {
+    OT_TCP_ACTIVE_SEND_MORE_TO_COME = 1 << 0,
+};
+
+/**
+ * Adds data referenced by the linked buffer chain pointed to by @p aBuffer to
+ * the send buffer. Upon a sucessful call to this function, the linked buffers
+ * and data they reference are owned by the TCP stack; they should not be
+ * modified by the application until a "send done" callback returns ownership
+ * of those objects to the application.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket    A pointer to the active TCP socket structure representing the TCP endpoint on which to send data.
+ * @param[in]  aBuffer    A pointer to the linked buffer chain referencing data to add to the send buffer.
+ * @param[in]  aFlags     Flags specifying options for this operation (see enumeration above).
+ *
+ * @retval OT_ERROR_NONE    Successfully added data to the send buffer.
+ * @retval OT_ERROR_FAILED  Failed to add data to the send buffer.
+ */
+otError otTcpActiveSendByReference(otInstance *aInstance, otTcpActiveSocket *aSocket, otLinkedBuffer *aBuffer, uint32_t aFlags);
```

Comment:
Considering that we need to copy the data from `otLinkedBuffer` to `otMessage` for sending the actual IPv6 datagram, how do we actually achieve `Zero Copy`?

Reply:
This is a good point. As I mentioned in our meeting, making a fully zero-copy solution would require modifying OpenThread to be able to send out packets represented directly as a linked buffer chain rather than an `otMessage`. This is an ambitious goal and it probably won't be implemented initially when we release TCP as part of OpenThread.

That said, I think it is valuable to design an API that's compatible with making TCP zero-copy. This will give us the flexibility to implement these zero-copy optimizations in the future. In contrast, if we stick to the simple `otTcpWrite` API, then we won't ever be able to make TCP zero-copy, because copying data is part of that API.

Reply:
> As I mentioned in our meeting, making a fully zero-copy solution would require modifying OpenThread to be able to send out packets represented directly as a linked buffer chain rather than an `otMessage`. This is an ambitious goal and it probably won't be implemented initially when we release TCP as part of OpenThread.

One possible way of making it happen is to use `otMessage` with just one inner `otMessageBuffer` that is large enough to hold the full TCP segment. We can then effectively use `otMessage` as the `Zero Copy` buffer and achieves real zero copying.

This would require TCP APIs to expose `otMessage` to users, like:

```
otMessage *otTcpNewMessage(void);
```

Then user can prepare data in the `otMessage` directly, which will be eventually sent as the IPv6 datagram. 

BTW, This strategy can also be used by #6036 to achieve `Zero Copy`.

Thoughts? @abtink @samkumar

Reply:
If I'm understanding this proposal correctly, then this would mean that the user has to be aware of TCP's segmentation (i.e., make sure to appropriately split the stream into separate `otMessage`s) right?

Personally, I feel that the policy to break up a stream into segments belongs in the TCP implementation, and the application code depend on how big the individual segments are. Basically, I don't think we should have the API requires the application developer to understand how to segment the stream into packets.

Reply:
To clarify, the `otLinkedBuffer`s in the proposed API don't correspond exactly to the packets being set out with TCP. For example, if each TCP packet is sized to include 462 bytes of data, each `otLinkedBuffer` in the send buffer could reference more than 462 bytes or less than 462 bytes.

To make TCP zero-copy in the long-term, we could modify OpenThread to allow IPv6 packets to be sent directly from `otLinkedBuffer` chains. Then the TCP stack could construct an `otLinkedBuffer` chain referencing data corresponding to segments in the send buffer to send them out. This could be done by allocating memory for new `otLinkedBuffer`s. Or, if we prefer to avoid allocating new `otLinkedBuffer`s, we could modify the `otLinkedBuffer` chain for the send buffer in-place to isolate individual segments to transmit them and then restore the original chain once transmission is complete.